docs/html/development/architecture/package-finding.rst

```diff
@@ -95,18 +98,47 @@ links.
 One of ``PackageFinder``'s main top-level methods is
 ``find_best_candidate()``. This method does the following two things:
 
-1. Calls its ``find_all_candidates()`` method, which reads and parses all the
-   index URL's provided by the user, constructs a :ref:`LinkEvaluator
-   <link-evaluator-class>` object to filter out some of those links, and then
-   returns a list of ``InstallationCandidates`` (aka candidates for install).
-   This corresponds to steps 1-3 of the :ref:`Overview <index-py-overview>`
-   above.
+1. Calls its ``find_all_candidates()`` method, which gathers all
+   possible package links by reading and parsing the index URL's and
+   locations provided by the user (the :ref:`LinkCollector
+   <link-collector-class>` class's ``collect_links()`` method), constructs a
+   :ref:`LinkEvaluator <link-evaluator-class>` object to filter out some of
+   those links, and then returns a list of ``InstallationCandidates`` (aka
+   candidates for install). This corresponds to steps 1-3 of the
+   :ref:`Overview <index-py-overview>` above.
 2. Constructs a ``CandidateEvaluator`` object and uses that to determine
    the best candidate. It does this by calling the ``CandidateEvaluator``
    class's ``compute_best_candidate()`` method on the return value of
    ``find_all_candidates()``. This corresponds to steps 4-5 of the Overview.
 
 
+.. _link-collector-class:
+
+The ``LinkCollector`` class
```

Comment:
I don't know if this discussion was had before, but would it be better to keep these kinds of docs in the code itself? Then we could just include it here in any number of ways and I think it would be more likely to be kept up to date.

Reply:
Thanks for taking a look! I don't really have a preference for how this type of documentation is done. The reason I added it in the first place is that there was a request from a number of people for me to document the various classes in a separate `rst` file, so I did that. And I was just updating it here. (I will note that one advantage of doing it directly in the `rst` is that it lets you use the various reStructuredText formatting, links, etc, whereas having that in a docstring would be a bit weird. I'm also not sure the extent to which it would carry over.)

Reply:
The original discussion was here: https://github.com/pypa/pip/pull/6787